### PR TITLE
Add back themehooks docs with additional examples

### LINKED ISSUE #201: Improve Theme hook documentation
https://docs.vanillaforums.com/developer/addons/events-and-handlers/

We should put some real examples in there for common use cases. I'd like to see:

- [x] A concrete and functional example of a theme hooks file for a theme
- [x] Show a few common theme hooks, especially `base_render_before`
- [x] Encouraging users to use theme hooks and not view overwrites (link from view overwrites back to theme hooks page)
- [x] Adding some data to `$sender`. It's important to be clear here that you can add what you want and offer a few different examples
- [x] Some mention or a link to other page with information about using `{debug}` to see and debug data being passed to view.
- [x] Links to smarty page(s)

I dislike using the wording "magic" and having abstract examples with names like "kaboom". Docs should be concrete and offer practical examples.

### PR DESCRIPTION
**Pull request by Charrondev**
Closes #201 

I've added back the old example and a lot of additional ones. I also refreshed most of the links. I moved the redirects from the old themehooks page that were on the events and handlers page to the new themehooks page.

### CODE REVIEW

File: content/developer/addons/events-and-handlers.md [hunk 1]
```diff
@@ -109,3 +106,188 @@ With this addon enabled, going to the URL `/discussions/kaboom` would now output
...
+Sometimes you may not need all the roles in the page. Let's say you wanted to make the page appear differently for a user with a certain role. Instead of injection all of the roles into the template and doing the conditional there, you can do it in the themehooks and inject just the boolean value you need.
+
+
+```php
+public function base_render_before($sender) {
+    if (inSection('Dashboard')) {
+        return;
+    }
+    
+    $roles = val('UserRoles', $sender->Data));
```
Comment (slafleche) [inline content/developer/addons/events-and-handlers.md]: Why is the code to fetch the roles different from the previous example?

File: content/developer/addons/events-and-handlers.md [hunk 2]
```diff
@@ -109,3 +106,188 @@ With this addon enabled, going to the URL `/discussions/kaboom` would now output
...
+     * Create the `/settings/example` page to host our custom settings.
+     *
+     * @param SettingsController $sender
+     */
+    public function settingsController_example_create($sender) {
+        $sender->permission('Garden.Settings.Manage');
+
+        $configurationModule = new ConfigurationModule($sender);
+        $configurationModule->initialize([
+            'ExmapleSite.BackgroundColour' => ['Control' => 'TextBox', 'Options' => ['class' => 'InputBox BigInput']],
```
Comment (slafleche) [inline content/developer/addons/events-and-handlers.md]: Do we have documentation anywhere on the types of inputs we accept? If we don't, we should make one. We should add a note with a link to that page.

File: content/developer/addons/theme-hooks.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+```
+
+
+## Common Hooks
+
+There are a few common functions for a themehooks file.
+
+### base_render_before()
+
+This runs on every single page load before the render. This includes the Dashboard so be sure to exclude it! This is also a good place to use the `Gdn_Controller` method `setData()` to inject data into your Smarty templates.
```
Comment (slafleche) [inline content/developer/addons/theme-hooks.md]: A link to `Gdn_Controller` and `setData()` would be nice here.

File: content/developer/addons/theme-hooks.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+### base_render_before()
+
+This runs on every single page load before the render. This includes the Dashboard so be sure to exclude it! This is also a good place to use the `Gdn_Controller` method `setData()` to inject data into your Smarty templates. 
+
+For information about using this injected data see [Accessing Controller Data with Smarty](/developer/smarty/#accessing-controller-data-with-smarty). If you're trying to verify that the data that is getting passed into the template, try out the [{debug} function](/developer/smarty/functions/#function-debug) in smarty.
+
+### setup()
+
+This runs when your theme is enabled. This is a good place for setting configuration values your theme may rely on, but is not required.
```
Comment (slafleche) [inline content/developer/addons/theme-hooks.md]: Add anchor to example at the end.

File: content/developer/addons/view-overrides.md [hunk 5]
```diff
@@ -123,7 +123,12 @@ To see how it all fits together, here's a default master view using Smarty:
...
+Now that you know how to override and configure the master view, you may want to configure the content of the assets and modules. Before we get there though, a warning: overriding a view can be a rather severe addition to a theme. Once a view is overridden, it diverges from core Vanilla. As such, **your new view may not always be supported** in future versions of Vanilla, which may cause problems down the road. 
+
+If possible first:
+
+- Try to accomplish what you're doing using good ol' CSS. A strong knowledge of CSS can allow you to make powerful customizations to your Vanilla Forum while being simpler, stabler, and more maintainable. The Vanilla Forums markup isn't perfect, but it's a lot easier down the road to tweak a few custom styles then to reconcile a totally custom view with the original if the shape of the data changes, or the view is refactored.
```
Comment (slafleche) [inline content/developer/addons/view-overrides.md]: Don't say our markup isn't perfect... I know it isn't, but that doesn't sound good ;P I would rather state that it's a smaller risk of the code breaking.

File: content/developer/addons/theme-hooks.md (hunk 4 above)
Comment (slafleche) [reply content/developer/addons/theme-hooks.md]: Also, explain structure() vs setup()

File: content/developer/addons/events-and-handlers.md (hunk 2 above)
Comment (Charrondev) [reply content/developer/addons/events-and-handlers.md]: We don't currently have proper documentation for Gdn_Form or the Configuration Module. These documents are a bit of a project in of themselves, so for now I'm going to link to the Gdn_Form implementation and some additional examples in vanilla where we use the ConfigurationModule.

I'll open an issue for the Gdn_Form docs though.

File: content/developer/addons/view-overrides.md (hunk 5 above)
Comment (Charrondev) [reply content/developer/addons/view-overrides.md]: I've adjusted this text.

File: content/developer/addons/events-and-handlers.md (hunk 1 above)
Comment (Charrondev) [reply content/developer/addons/events-and-handlers.md]: Corrected.

File: content/developer/addons/theme-hooks.md (hunk 3 above)
Comment (Charrondev) [reply content/developer/addons/theme-hooks.md]: I've added these to the Controller documentation and linked to them.

File: content/developer/addons/events-and-handlers.md (hunk 2 above)
Comment (Charrondev) [reply content/developer/addons/events-and-handlers.md]: Additionally I've added a list of the support form components in the ConfigurationModule

**Pull request merged**